Moist-Dry conversions

jatkinson1000 opened the pull request:
This addresses #15 

The variable conversion routines now also include a moist - dry conversion and back.

The moist mixing ratios from CAM are now converted to dry mixing ratios as expected by SAM for use in that code, and then back for calculation of tendencies and return values.

jatkinson1000 committed:
Update CAM interface to YOG to convert moist mix. ratios to dry for use in the SAM parts of the code.

jatkinson1000 committed:
Update CAM interface to YOG to convert dry mix. ratios to moist for use in the returning CAM parts of the code.

MarionBWeinzierl commented on src/physics/cam/nn_interface_cam.F90:
```diff
@@ -123,8 +123,8 @@ subroutine nn_convection_flux_CAM(pres_cam, pres_int_cam, pres_sfc_cam, &
             !! precipitation at surface from one call to parameterisation
 
         ! Local input variables on the SAM grid
```
Personally, I would find some info in the comments useful what these variables are. Or is that info available in documentation elsewhere?

jatkinson1000 replied:
I have fleshed out the comments and documentation in https://github.com/m2lines/CAM-ML/pull/20/commits/498d6fdfc6ab7ef5365036b918a0ef0854188b56

MarionBWeinzierl reviewed (approved):
This looks sensible to me. Some more comments might help, in particular what the variables are and what happens, e.g., in the iteration in lines 687-716. Otherwise fine.

jatkinson1000 committed:
Add comments to explain process of variable conversion when iterating for condensation.

jatkinson1000 commented:
I have added some comments and references to explain the variable conversion code in https://github.com/m2lines/CAM-ML/pull/20/commits/f4e7198f641b435c2835774225f4b3dfa8217404

jatkinson1000 committed:
Improve documentation for the various variables in the nn_flux routine as requested by @ MarionBWeinzierl.

jatkinson1000 committed:
Set array size of chunked precipitation to be ncols instead of pcols.

paogorman reviewed (commented):
I'm a bit worried about the introduction of the r variables for mixing ratios. We have ended up with lines like: 
 qn = max(0., r_temp-qsat)
Here qn, qsat, and r_temp are all mixing ratios but only one is denoted r.
An alternative approach would be to keep everything as q, and have distinct blocks of code that do the moist to dry mixing ratio conversions.

MarionBWeinzierl reviewed (commented):
I am happy with this now, but will refrain from approving it until @paogorman is happy, too

jatkinson1000 committed:
rename q to r when relating to SAM as CAM uses q for moist mixing and SAM is dry mixing.

jatkinson1000 commented:
I have renamed the variables used by 'SAM' to be `r` instead of `q`, and renamed the functions extracted from SAM to use `r` in their names as well.
If you can review and check you are happy before I merge it that would be appreciated.

paogorman reviewed (commented):
Change to rv etc. generally looks good and does make things clearer.
For CAM_var_conversion:
- might be cleaner to introduce rc,ri explicitly (but what you have looks right)
- are you now missing a declaration for incoming variables qc,qi?

jatkinson1000 commented on src/physics/cam/nn_interface_cam.F90:
```diff
@@ -593,28 +603,30 @@ subroutine CAM_var_conversion(qv, qc, qi, q, tabs, t)
+            ! Calculate dry mixing ratio as required by SAM from moist variables as provided by CAM
+            rv = qv(i,k) / (1. - qv(i,k))
+            r(i,k) = rv + qc(i,k)*(1.+rv) + qi(i,k)*(1.+rv)
 
             ! omp  = max(0.,min(1.,(tabs(i,k)-tprmin)*a_pr))
             omn  = max(0.,min(1.,(tabs(i,k)-tbgmin)*a_bg))
             t(i,k) = tabs(i,k) &
                    ! - (fac_cond+(1.-omp)*fac_fus)*qp(i,k) &  ! There is no qp in CAM
-                     - (fac_cond+(1.-omn)*fac_fus)*(qc(i,k) + qi(i,k)) &
+                     - (fac_cond+(1.-omn)*fac_fus)*(qc(i,k)*(1.+rv) + qi(i,k)*(1.+rv)) &
                      + gamaz(k)
           end do
         end do
 
     end subroutine CAM_var_conversion
     
```
Move conversion to rc and ri out of ll. 619 to be below 613, and create an explicit ri and rc variable.

jatkinson1000 replied:
This should now be done in https://github.com/m2lines/CAM-ML/pull/20/commits/7ceba23360ed728223c79989c9a5413fe3d0cadf (with a further simplification made in calculating `r(i,k)`).

jatkinson1000 committed:
Add intermediate variables for dry mixing ratios of liq and ice in CAM->SAM variable conversion to improve clarity.

jatkinson1000 commented:
This should now be ready to go.
@paogorman the changes we discussed are summarised in https://github.com/m2lines/CAM-ML/pull/20/commits/15e21ccaa9ee6233889ebb96882b5a9b4765012a if you could have a quick glance to check you are happy with that. I also used it to simplify the calculation of `r(i,k)` as well, as this also uses `ri` and `rc` so avoids a double calculation.

paogorman commented:
Yes this looks good - thanks Jack
